### ISSUE TITLE
SslStream doesn't update its stored buffer, causing excess allocations

### ISSUE THREAD

**Issue by Porges** [dotnet/corefx#23040]
[Here in `SslStream`](https://github.com/dotnet/corefx/blob/cb173f6004e9671793d1aff64bab78e2d5f701c3/src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs#L419), `EncryptData` can update the passed `outBuffer`. However, since SslStream throws away this value and doesn't update the fields in the class, this can cause excessive allocations to occur. (The allocation itself occurs in [`SslStreamPal`](https://github.com/dotnet/corefx/blob/cb173f6004e9671793d1aff64bab78e2d5f701c3/src/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs#L149)).

At the moment I'm working around this by writing to the stream with very specifically-sized buffers (i.e. 4096-bytes), which means that the new allocation doesn't occur.

Here's a little snapshot from a DotMemory session showing this behaviour (the numbers are bytes allocated, instances allocated, bytes collected, instances collected):

![image](https://user-images.githubusercontent.com/12575/29059530-56e11508-7c59-11e7-8b47-51f9e79781f2.png)

**Comment by karelz** [comment]
cc @Drawaes

**Comment by Drawaes** [comment]
I agree this causes excessive allocations. However just updating the field will cause more problems than it solves. The field currently is allocated from the Pinnable Buffer Pool. If you update the field you are now holding that array instead of the pooled one.

There are really two options here to fix the issue at hand. Do the chunking at the same size as the write buffer, to stop extra allocations.
Increase the writebuffer size to the max frame size plus header for TLS (2^14 + 5 + 32).

With the current design however that means that every connection will now hold an extra 11k that can't/won't be released. In a scenario where a server has a lot of idle connections this can add even more overhead per connection (image say StackOverflow with 60k or more websockets lying around not doing much). 

Already in OpenSsl there is idle buffers in the Ssl object, the Memory Bio, and the read and write pinned buffers.

So to sum up, I agree this should be fixed, and if you take a look at the flow has been changed
#22998

So there is with that change a renting of the pooled buffer for only the duration of the encrypt/write phase. This means you could happily then increase the buffer size to the framesize and not worry about eating your memory. I haven't optimized the Windows or OSX side yet to take advantage of this, but you can see the results for the *nix side in my graph (to be fair there are a couple of other improvements like removing that extra memory bio).

Cheers Tim

**Comment by Drawaes** [comment]
Just as a final note... (not on windows cause I haven't hit that side much yet) but on the *nix side I am getting zero allocations (in fact it is not possible anymore) for that write buffer, as I am using OpenSsl's native ability to write out 5k of the frame at a time.

**Comment by Drawaes** [comment]
@Porges This should have been fixed by the above PR that is now merged. There is a buffer sized for your current request/chunk size so it should never allocate (other than to fill the ArrayPool). The code to make the new allocated buffer is still in the PAL but should never be needed. I have a PR prepared just waiting for some other PR's to clear that removes that now defunct code.

Cheers
Tim

**Comment by karelz** [comment]
Closing the issue as fixed per @Drawaes's comment above.
cc @geoffkizer

**Comment by Porges** [comment]
Thanks! 😊
